joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/scope/Scope.scala

```diff
@@ -0,0 +1,270 @@
+package io.joern.javasrc2cpg.scope
+
+import io.shiftleft.codepropertygraph.generated.nodes.NewTypeDecl
+import io.shiftleft.codepropertygraph.generated.nodes.NewNamespace
+import io.shiftleft.codepropertygraph.generated.nodes.NewMethod
+import io.shiftleft.codepropertygraph.generated.nodes.NewBlock
+import io.shiftleft.codepropertygraph.generated.nodes.NewTypeParameter
+import io.shiftleft.codepropertygraph.generated.nodes.NewImport
+import io.shiftleft.codepropertygraph.generated.nodes.NewLocal
+import io.shiftleft.codepropertygraph.generated.nodes.NewMethodParameterIn
+import io.shiftleft.codepropertygraph.generated.nodes.NewMember
+import io.shiftleft.codepropertygraph.generated.nodes.NewNode
+import io.shiftleft.codepropertygraph.generated.nodes.DeclarationNew
+
+import io.joern.javasrc2cpg.scope.Scope._
+import io.joern.javasrc2cpg.passes.ExpectedType
+
+import scala.collection.mutable
+import io.joern.x2cpg.utils.ListUtils._
+import io.shiftleft.codepropertygraph.generated.nodes.NewNamespaceBlock
+import io.joern.x2cpg.Ast
+import io.joern.javasrc2cpg.util.NameConstants
+
+// TODO: Added for backwards compatibility with old scope methods, but is no longer
+//  strictly necessary due to stricter scope variable classes. Refactor AstCreator
+//  to make use of those and remove this instead.
+case class NodeTypeInfo(
+  node: NewNode,
+  name: String,
+  typeFullName: Option[String],
+  isField: Boolean = false,
+  isStatic: Boolean = false
+)
+class Scope {
+  private var scopeStack: List[JavaScopeElement] = Nil
+
+  def pushBlockScope(): Unit = {
+    scopeStack = new BlockScope() :: scopeStack
+  }
+
+  def pushMethodScope(method: NewMethod, returnType: ExpectedType): Unit = {
+    scopeStack = new MethodScope(method, returnType) :: scopeStack
+  }
+
+  def pushTypeDeclScope(typeDecl: NewTypeDecl): Unit = {
+    scopeStack = new TypeDeclScope(typeDecl) :: scopeStack
+  }
+
+  def pushNamespaceScope(namespace: NewNamespaceBlock): Unit = {
+    scopeStack = new NamespaceScope(namespace) :: scopeStack
+  }
+
+  def popScope(): JavaScopeElement = {
+    scopeStack.headOption match {
```

Comment:
I would just to `val scope = scopeStack.head;scopeStack = scopeStack.tail;scope`. this throws an exception if the stack is empty like you also do at the moment. The custom text message is not required.
Please apply the same to the `addXXX` methods.